embedded-service/src/ec_type/mod.rs

```diff
@@ -1,4 +1,1123 @@
+        message::ThermalMessage::Fan1HotTemp(fan1_hot_temp) => memory_map.therm.fan1_hot_temp = *fan1_hot_temp,
+        message::ThermalMessage::Fan1MaxRpm(fan1_max_rpm) => memory_map.therm.fan1_max_rpm = *fan1_max_rpm,
+        message::ThermalMessage::Fan1CurRpm(fan1_cur_rpm) => memory_map.therm.fan1_cur_rpm = *fan1_cur_rpm,
+        message::ThermalMessage::Tmp1Val(tmp1_val) => memory_map.therm.tmp1_val = *tmp1_val,
+        message::ThermalMessage::Tmp1Timeout(tmp1_timeout) => memory_map.therm.tmp1_timeout = *tmp1_timeout,
+        message::ThermalMessage::Tmp1Low(tmp1_low) => memory_map.therm.tmp1_low = *tmp1_low,
+        message::ThermalMessage::Tmp1High(tmp1_high) => memory_map.therm.tmp1_high = *tmp1_high,
+    }
+}
+
+/// Update time alarm section of memory map based on battery message
+pub fn update_time_alarm_section(msg: &message::TimeAlarmMessage, memory_map: &mut structure::ECMemory) {
+    match msg {
+        message::TimeAlarmMessage::Events(events) => memory_map.alarm.events = *events,
+        message::TimeAlarmMessage::Capability(capability) => memory_map.alarm.capability = *capability,
+        message::TimeAlarmMessage::Year(year) => memory_map.alarm.year = *year,
+        message::TimeAlarmMessage::Month(month) => memory_map.alarm.month = *month,
+        message::TimeAlarmMessage::Day(day) => memory_map.alarm.day = *day,
+        message::TimeAlarmMessage::Hour(hour) => memory_map.alarm.hour = *hour,
+        message::TimeAlarmMessage::Minute(minute) => memory_map.alarm.minute = *minute,
+        message::TimeAlarmMessage::Second(second) => memory_map.alarm.second = *second,
+        message::TimeAlarmMessage::Valid(valid) => memory_map.alarm.valid = *valid,
+        message::TimeAlarmMessage::Daylight(daylight) => memory_map.alarm.daylight = *daylight,
+        message::TimeAlarmMessage::Res1(res1) => memory_map.alarm.res1 = *res1,
+        message::TimeAlarmMessage::Milli(milli) => memory_map.alarm.milli = *milli,
+        message::TimeAlarmMessage::TimeZone(time_zone) => memory_map.alarm.time_zone = *time_zone,
+        message::TimeAlarmMessage::Res2(res2) => memory_map.alarm.res2 = *res2,
+        message::TimeAlarmMessage::AlarmStatus(alarm_status) => memory_map.alarm.alarm_status = *alarm_status,
+        message::TimeAlarmMessage::AcTimeVal(ac_time_val) => memory_map.alarm.ac_time_val = *ac_time_val,
+        message::TimeAlarmMessage::DcTimeVal(dc_time_val) => memory_map.alarm.dc_time_val = *dc_time_val,
+    }
+}
+
+/// Convert from memory map offset and length to battery message
+/// Modifies offset and length
+pub fn mem_map_to_battery_msg(
+    memory_map: &structure::ECMemory,
+    offset: &mut usize,
+    length: &mut usize,
+) -> Result<message::BatteryMessage, Error> {
+    let local_offset = *offset - offset_of!(structure::ECMemory, batt);
+    let mut message: Option<message::BatteryMessage> = None;
```

Comment:
this can be made immutable if you make the `if` that follows evaluate to the message. It would look something like:

```rust
let message = match local_offset {
    offset_of!(structure::Battery, events) => {
        // ...
    }

    // ...

    _ => None
};
```

But if you do that, then it makes sense to do away with the `Option<T>` and just evaluate to the `Result<T, E>` at the end, thus removing the local variable:

```rust
match local_offset {
    offset_of!(structure::Battery, events) => {
        // ...
        Some(message)
    }

    // ...

    _ => Err(Error::InvalidLocation)
}
```